File: Source/Aladdin.CAPI.ANSI.PKCS11/RSA/PublicKey.cs
using System;
using Aladdin.PKCS11; 

namespace Aladdin.CAPI.ANSI.PKCS11.RSA
{
    ///////////////////////////////////////////////////////////////////////////
    // Открытый ключ RSA
    ///////////////////////////////////////////////////////////////////////////
    public class PublicKey : CAPI.PublicKey, CAPI.ANSI.RSA.IPublicKey
    {
        // способ кодирования чисел
        private const Math.Endian Endian = Math.Endian.BigEndian; 
    
	    private Math.BigInteger modulus;        // параметр N
	    private Math.BigInteger publicExponent;	// параметр E
    
	    // атрибуты открытого ключа
	    public static CAPI.PKCS11.Attribute[] GetAttributes(
            CAPI.PKCS11.Provider provider, CAPI.ANSI.RSA.IPublicKey publicKey)
        {
            // закодировать параметры открытого ключа
            byte[] modulus        = Math.Convert.FromBigInteger(publicKey.Modulus       , Endian); 
            byte[] publicExponent = Math.Convert.FromBigInteger(publicKey.PublicExponent, Endian);
        
            // создать набор атрибутов
            return new CAPI.PKCS11.Attribute[] { provider.CreateAttribute(API.CKA_KEY_TYPE, API.CKK_RSA),

                // указать идентификаторы параметров
                provider.CreateAttribute(API.CKA_MODULUS        , modulus       ),
                provider.CreateAttribute(API.CKA_PUBLIC_EXPONENT, publicExponent) 
            }; 
        }
        // конструктор
	    public PublicKey(CAPI.PKCS11.Provider provider, CAPI.PKCS11.SessionObject obj)

            // сохранить переданные параметры
            : base(provider.GetKeyFactory(ASN1.ISO.PKCS.PKCS1.OID.rsa))
        {
		    // получить атрибуты ключа
		    CAPI.PKCS11.Attributes keyAttributes = provider.GetKeyAttributes(obj, 
                new CAPI.PKCS11.Attribute(API.CKA_MODULUS        ), 
                new CAPI.PKCS11.Attribute(API.CKA_PUBLIC_EXPONENT) 
            ); 
            // раскодировать значение параметров
            modulus = Math.Convert.ToBigInteger(
                keyAttributes[API.CKA_MODULUS].Value, Endian
            );
            publicExponent = Math.Convert.ToBigInteger(
                keyAttributes[API.CKA_PUBLIC_EXPONENT].Value, Endian
            );
        }
        // параметры ключа
        public override IParameters Parameters
        {
            // параметры ключа
            get { return new ANSI.RSA.Parameters(modulus.BitLength, publicExponent); }     
        }
	    public Math.BigInteger Modulus        { get { return modulus;        }}
	    public Math.BigInteger PublicExponent { get { return publicExponent; }}
    }
}

File: Source/Aladdin.CAPI.ANSI.PKCS11.Java/src/aladdin/capi/ansi/pkcs11/rsa/PublicKey.java
package aladdin.capi.ansi.pkcs11.rsa;
import aladdin.math.*; 
import aladdin.capi.*;
import aladdin.capi.pkcs11.*;
import aladdin.pkcs11.*;
import java.math.*;
import java.io.*; 

///////////////////////////////////////////////////////////////////////////
// Открытый ключ RSA
///////////////////////////////////////////////////////////////////////////
public class PublicKey extends aladdin.capi.PublicKey implements aladdin.capi.ansi.rsa.IPublicKey
{
    // способ кодирования чисел
    private static final Endian ENDIAN = Endian.BIG_ENDIAN; 
    
	private final BigInteger modulus;           // параметр N
	private final BigInteger publicExponent;	// параметр E
    
	// атрибуты открытого ключа
	public static Attribute[] getAttributes(aladdin.capi.ansi.rsa.IPublicKey publicKey)
    {
        // закодировать параметры открытого ключа
        byte[] modulus        = Convert.fromBigInteger(publicKey.getModulus       (), ENDIAN); 
        byte[] publicExponent = Convert.fromBigInteger(publicKey.getPublicExponent(), ENDIAN);
        
        // создать набор атрибутов
        return new Attribute[] { new Attribute(API.CKA_KEY_TYPE, API.CKK_RSA),

            // указать идентификаторы параметров
            new Attribute(API.CKA_MODULUS        , modulus       ),
            new Attribute(API.CKA_PUBLIC_EXPONENT, publicExponent) 
        }; 
    }
    // конструктор
	public PublicKey(aladdin.capi.pkcs11.Provider provider, SessionObject object) throws IOException
    {
        // сохранить переданные параметры
        super(provider.getKeyFactory(aladdin.asn1.iso.pkcs.pkcs1.OID.RSA)); 
        
		// получить атрибуты ключа
		Attributes keyAttributes = provider.getKeyAttributes(object, 
            new Attribute(API.CKA_MODULUS        , byte[].class), 
            new Attribute(API.CKA_PUBLIC_EXPONENT, byte[].class) 
        ); 
        // раскодировать значение параметров
        modulus = Convert.toBigInteger(
            (byte[])keyAttributes.get(API.CKA_MODULUS).value(), ENDIAN
        );
        publicExponent = Convert.toBigInteger(
            (byte[])keyAttributes.get(API.CKA_PUBLIC_EXPONENT).value(), ENDIAN
        );
    }
    // параметры ключа
    @Override public IParameters parameters()
    {
        // параметры ключа
        return new aladdin.capi.ansi.rsa.Parameters(modulus.bitLength(), publicExponent);     
    }
	@Override public final BigInteger getModulus        () { return modulus;        }
	@Override public final BigInteger getPublicExponent () { return publicExponent;	}
}
